Provide entry point similar to builtin functions for both dynamic calls and user-defined function calls.

ghislainfourny opened the pull request.

CanBerker commented on src/main/java/sparksoniq/jsoniq/runtime/iterator/functions/FunctionItemCallIterator.java:
```diff
@@ -37,11 +37,10 @@
 
     private static final long serialVersionUID = 1L;
     // parametrized fields
-    private FunctionIdentifier _functionIdentifier;
```
Can you move _functionitem(defined below) to thisline to keep consistency with the comments regarding parametrized fields or calculated fields. If these comments are not helpful, you can also remove them if you wish

ghislainfourny replied:
Done! Good catch. The comments make sense.

CanBerker commented on src/main/java/sparksoniq/jsoniq/runtime/iterator/functions/base/Functions.java:
```diff
@@ -292,6 +293,22 @@ public static RuntimeIterator getBuiltInFunctionIterator(
+            IteratorMetadata metadata,
+            List<RuntimeIterator> arguments
+    ) {
+        return buildUserDefinedFunctionIterator(getUserDefinedFunction(identifier, metadata), metadata, arguments);
```
I think getUserDefinedFunction(identifier, metadata) can be simplified to not have the metadata parameter. It seems unused

ghislainfourny replied:
Also good catch, done.

CanBerker commented on src/main/java/sparksoniq/jsoniq/runtime/iterator/functions/DynamicallyResolvedFunctionCallIterator.java:
```diff
@@ -0,0 +1,147 @@
+            _functionIdentifier,
+            getMetadata(),
+            _functionArguments
+        );
```
If I'm not mistaken the assignment to _functionCalliterator is redundant. This lookup is already performed in initIsRdd (which is always called before openLocal() or getRDDAux() ) method and saved to class state.

ghislainfourny replied:
Good catch, thanks.

Note to self: resolving the function at init time may prevent recursive call. We may need to move it back to open and getRDD later on to enable recursion.

ghislainfourny commented on src/main/java/sparksoniq/jsoniq/runtime/iterator/functions/DynamicFunctionCallIterator.java:
```diff
@@ -73,82 +74,10 @@ public DynamicFunctionCallIterator(
     @Override
     public void openLocal() {
         setFunctionItemAndIteratorWithCurrentContext();
```
@CanBerker does this apply here too that it is redundant?

CanBerker replied:
Precisely!

ghislainfourny replied:
Very good. For dynamic calls recursion is not an issue as I do not think it is (as of the current version of the W3C spec) possible for a function item to call itself.

ghislainfourny replied:
I reverted, it seems that it does make a difference. We could investigate later on.

ghislainfourny replied:
I think I found the issue. It is because there was this partial application test, but it is redundant with the other iterator. Now it should work.

ghislainfourny replied:
I put it back for now -- there seems to be an issue with serializing and deserializing function items that forces us to reset the function when opening, which we can investigate later.

CanBerker commented:
The PR looks great overall. I created a new PR #378 (lined up to merge into this one) to discuss the naming of the new classes. If you wish, we can directly merge this PR too, I'll be waiting on your feedback to proceed.

ghislainfourny commented:
@CanBerker one more fix; the dynamic function call iterator must rebuild the function item iterator every time, because it is dynamic-context-dependent. I added a test that was broken and now fixed.

https://github.com/RumbleDB/rumble/pull/377/commits/a3791ee12793fe8f845d627d3405aa1791161a0c
